Add/block product categories

ryelle opened the pull request:
Fixes #212  – Add a new block for a Product Categories list. This is using another approach to rendering, using the JS on the front-end as well. I've put up this draft PR to get feedback on this as an approach, since I think we'll want to do something similar for the 2 product reviews blocks (#588 & #589).

Currently you can add the block, and it renders on the frontend, but you can't reload or re-open the editor with an existing Product Categories List block– something's breaking the attribute validation.

**Questions**

- Does the folder & component structure make sense?
- Should we attempt to render any content into the `save` function? (loading, some skeleton, ?)
- Thoughts on this approach in general?

For now, just cc'ing @woocommerce/blocks – will request proper reviews once this is ready.

Still to do:

- Fix whatever bug is preventing loading an existing block
- Add links to the links
- Add navigation code to the dropdown in an accessible way
- Make sure nav only happens on frontend
- Add hierarchy option to dropdown view

#### Accessibility

Before un-drafting I'll update this section.

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

![Screen Shot 2019-05-30 at 5 37 20 PM](https://user-images.githubusercontent.com/541093/58666567-afd9c000-8301-11e9-9d40-d94d63bee27d.png)
![Screen Shot 2019-05-30 at 5 37 09 PM](https://user-images.githubusercontent.com/541093/58666568-b0725680-8301-11e9-944f-a7fa4bac67dd.png)
![Screen Shot 2019-05-30 at 5 36 53 PM](https://user-images.githubusercontent.com/541093/58666569-b0725680-8301-11e9-96ed-889f6eac2bb6.png)

### How to test the changes in this Pull Request:

1. Add a Product Categories List block
2. You should be able to toggle all the toggles, and have that reflect on the preview and in the frontend.

timmyc commented on assets/php/class-wgpb-block-library.php:
```diff
@@ -354,6 +377,14 @@ public static function print_script_settings() {
+		$product_categories = get_terms(
+			'product_cat',
+			array(
+				'hide_empty' => false,
```
Any reason behind adding in these two arguments here? Guessing get_terms can be filtered to over-ride, just more curious really.

ryelle replied:
`hide_empty` so that we always get all terms, `pad_counts`… I don't remember. WC core used it so I copied it 😅

ryelle commented on assets/js/blocks/product-categories/block.js:
```diff
@@ -0,0 +1,77 @@
+	/**
+	 * The attributes for this block
+	 */
+	attributes: PropTypes.object.isRequired,
```
Those are properties of `attributes` – we don't usually document those since they're set up in the `registerBlockType` call

ryelle commented:
> The approach looks good to me. Locally I'm not seeing any block render on the front-end... though also looking for wc-frontend js file in the dom and night finding anything so that might be the issue.

@timmyc Hm, you should be seeing it… do you see it in your `build` folder?

ryelle commented:
I've updated the PR & rebased onto #626 (so we can handle dependencies better). I also fixed the issue with reloading an existing post with this block, so that should be working.

Currently the loading placeholder is _very much_ a placeholder, but I think this is ready for a dev review, and we can add a proper loading placeholder in a following PR.

Rahmon commented on assets/js/blocks/product-categories/block.js:
```diff
@@ -0,0 +1,70 @@
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { noop } from 'lodash';
```
What do you think import this function like that `import noop from 'lodash/noop'` ?

References:
https://www.blazemeter.com/blog/the-correct-way-to-import-lodash-libraries-a-benchmark/

ryelle replied:
This is rewritten when built to reference the lodash that's already on the page, as part of WordPress core (using [`@wordpress/dependency-extraction-webpack-plugin`](https://developer.wordpress.org/block-editor/packages/packages-dependency-extraction-webpack-plugin/)). Since `lodash` is already on the page (and not in our bundle), we don't gain anything by importing like that.

joshuatf commented on assets/js/blocks/product-categories/index.js:
```diff
@@ -0,0 +1,98 @@
+		if ( isHierarchical ) {
+			props[ 'data-is-hierarchical' ] = true;
+		}
+		return <div { ...props }>LOADING</div>;
```
My only concern with this is without any SSR for categories, will this affect SEO for sites?

ryelle replied:
I don't think so– search engines know about JS. It seems that google, at least, will wait about 5 seconds after the load event to index the page, so the categories should be loaded by then.

joshuatf commented on assets/js/blocks/product-categories/frontend.js:
```diff
@@ -0,0 +1,27 @@
+);
+
+if ( containers.length ) {
+	containers.forEach( ( el ) => {
```
This a nifty way of handling this 👍 

Obviously beyond the scope of this PR, but if this does become the standard for frontend blocks, it would be cool to have a method that can automatically detect and pull attributes from a given class for any given block.

ryelle commented:
> In the editor you can click the dropdown nav (it shows categories but does nothing else). This could be done accidentally when selecting the block. Should interaction be disabled?

I think this is OK, since you want to be able to open the dropdown to check which categories are displayed.

> On the frontend if you view source you see LOADING. Is this how it will be shipped?

Nope, I just wanted to get this PR reviewed before I kept tweaking the placeholder.

Thanks for all the reviews! I'm going to merge this one and address the rest of my todos in following PRs 👍
